core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -780,6 +782,24 @@ object IO {
     case x :: xs => x.fork *> forkAll(xs)
   }
 
+  /**
+    * Apply the function fn to each element of the `TraversableOnce[A]` and
+    * return the results in a new `TraversableOnce[B]`.
+    */
+  def traverse[E, A, B, M[X] <: TraversableOnce[X]](in: M[A])(fn: A => IO[E, B])(
+    implicit cbf: CanBuildFrom[M[A], B, M[B]]): IO[E, M[B]] = {
+
+    def loopList(l: List[A], b: mutable.Builder[B, M[B]]): IO[E, M[B]] = l match {
+      case Nil =>
+        now(b.result())
+      case h :: t =>
+        fn(h).flatMap(a => loopList(t, b += a))
+    }
+
+    loopList(in.toList, cbf(in))
```

Comment:
You don't need to convert it to list. Also it would be more idiomatic to use `foldLeft` from `TraversableOnce` to do the looping instead of raw recursion.

Reply:
`foldLeft` is implemented in terms of `foreach` which is basically a `while` loop, so it will be easy on the stack :smiley:
If you get stuck you can use the implementation of `Future#traverse` as a reference.